File: andriod/app/src/main/java/com/Bands70k/FirebaseConnectionHelper.java
package com.Bands70k;

import android.os.Handler;
import android.os.Looper;
import android.util.Log;

import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;

import java.io.BufferedReader;
import java.io.FileReader;

/**
 * Lazily opens and closes Firebase Realtime Database connections to stay within concurrent connection limits.
 *
 * One WebSocket per app. Sessions are refcounted so overlapping user/band/show writes share one
 * connection, then disconnect after a short idle so a herd of launches does not stay online.
 */
public final class FirebaseConnectionHelper {
    private static final String TAG = "FirebaseConnectionHelper";
    private static final Object LOCK = new Object();
    private static final Handler HANDLER = new Handler(Looper.getMainLooper());
    private static final long IDLE_CLOSE_DELAY_MS = 1000L;
    public static final int LAUNCH_JITTER_MAX_MS = 20_000;

    private static int writeSessionCount = 0;
    private static Runnable idleCloseRunnable;

    private FirebaseConnectionHelper() {}

    public static DatabaseReference databaseReference() {
        return FirebaseDatabase.getInstance().getReference();
    }

    /**
     * Opens the RTDB connection if this is the first in-flight write on this device.
     */
    public static DatabaseReference beginWriteSession(String reason) {
        synchronized (LOCK) {
            cancelIdleCloseLocked();
            writeSessionCount++;
            if (writeSessionCount == 1) {
                try {
                    FirebaseDatabase.getInstance().goOnline();
                    Log.d(TAG, "goOnline (" + reason + ") sessions=1");
                } catch (Exception error) {
                    writeSessionCount = Math.max(0, writeSessionCount - 1);
                    Log.w(TAG, "goOnline failed (" + reason + "): " + error.getMessage());
                    return null;
                }
            } else {
                Log.d(TAG, "reuse connection (" + reason + ") sessions=" + writeSessionCount);
            }
        }
        return FirebaseDatabase.getInstance().getReference();
    }

    /**
     * Closes the RTDB connection when the last in-flight write finishes (after a 1s idle).
     */
    public static void endWriteSession(String reason) {
        synchronized (LOCK) {
            writeSessionCount = Math.max(0, writeSessionCount - 1);
            Log.d(TAG, "endWriteSession (" + reason + ") sessions=" + writeSessionCount);
            if (writeSessionCount > 0) {
                return;
            }
            cancelIdleCloseLocked();
            idleCloseRunnable = () -> {
                synchronized (LOCK) {
                    if (writeSessionCount == 0) {
                        try {
                            FirebaseDatabase.getInstance().goOffline();
                            Log.d(TAG, "goOffline (idle after " + reason + ")");
                        } catch (Exception error) {
                            Log.w(TAG, "goOffline failed (" + reason + "): " + error.getMessage());
                        }
                    }
                }
            };
            HANDLER.postDelayed(idleCloseRunnable, IDLE_CLOSE_DELAY_MS);
        }
    }

    /**
     * After Firebase auto-init the SDK is online. Disconnect until a write session starts.
     */
    public static void goIdleAfterConfigure() {
        synchronized (LOCK) {
            if (writeSessionCount != 0) {
                return;
            }
            try {
                FirebaseDatabase.getInstance().goOffline();
                Log.d(TAG, "goOffline (post-configure idle)");
            } catch (Exception error) {
                Log.w(TAG, "goOffline post-configure failed: " + error.getMessage());
            }
        }
    }

    private static void cancelIdleCloseLocked() {
        if (idleCloseRunnable != null) {
            HANDLER.removeCallbacks(idleCloseRunnable);
            idleCloseRunnable = null;
        }
    }

    /** Spreads connection opens across launches using a stable per-device delay (0–20s). */
    public static int jitterDelayMs(String userId, int maxJitterMs) {
        if (userId == null || userId.isEmpty()) {
            return 0;
        }
        return Math.floorMod(userId.hashCode(), maxJitterMs + 1);
    }

    /**
     * Festival year for Firebase paths: always {@code Current::eventYear} from the pointer cache,
     * never UI browse year or calendar fallback.
     */
    public static int firebaseStorageEventYear() {
        int fromFile = readCurrentEventYearFromPointerCacheFile();
        if (fromFile > 2000) {
            return fromFile;
        }

        staticVariables.loadUrlsFromCachedPointerFile("Current");
        if (staticVariables.storePointerData != null) {
            String yearStr = staticVariables.storePointerData.get("eventYear");
            if (yearStr != null && !yearStr.trim().isEmpty()) {
                try {
                    int year = Integer.parseInt(yearStr.trim());
                    if (year > 2000) {
                        return year;
                    }
                } catch (NumberFormatException ignored) {
                }
            }
        }

        Log.e(TAG, "Unable to resolve pointer Current event year for Firebase storage");
        return 0;
    }

    private static int readCurrentEventYearFromPointerCacheFile() {
        if (!FileHandler70k.pointerCacheFile.exists()) {
            return 0;
        }
        try (BufferedReader reader = new BufferedReader(new FileReader(FileHandler70k.pointerCacheFile))) {
            String line;
            while ((line = reader.readLine()) != null) {
                String trimmed = line.trim();
                if (!trimmed.startsWith("Current::eventYear::")) {
                    continue;
                }
                String[] parts = trimmed.split("::");
                if (parts.length >= 3) {
                    return Integer.parseInt(parts[2].trim());
                }
            }
        } catch (Exception error) {
            Log.w(TAG, "Failed reading Current event year from pointer cache: " + error.getMessage());
        }
        return 0;
    }
}

File: swift/70000TonsBands/FirebaseConnectionHelper.swift
//
//  FirebaseConnectionHelper.swift
//  70K Bands
//

import Foundation
import Firebase

/// Lazily opens and closes Firebase Realtime Database connections to stay within concurrent connection limits.
///
/// One WebSocket per app. Sessions are refcounted so overlapping user/band/show writes share one
/// connection, then disconnect after a short idle so a herd of launches does not stay online.
enum FirebaseConnectionHelper {
    private static let queue = DispatchQueue(label: "FirebaseConnectionHelper.queue")
    private static var writeSessionCount = 0
    private static var idleCloseWork: DispatchWorkItem?
    /// Coalesce back-to-back writes (user then band) into one connection pulse.
    private static let idleCloseDelaySeconds: TimeInterval = 1.0
    static let launchJitterMaxMs = 20_000

    static func databaseReference() -> DatabaseReference? {
        guard AppDelegate.isFirebaseConfigured else {
            print("⚠️ [FIREBASE_CONN] Firebase not configured — skipping database reference")
            return nil
        }
        return Database.database().reference()
    }

    /// Opens the RTDB connection if this is the first in-flight write on this device.
    @discardableResult
    static func beginWriteSession(reason: String) -> DatabaseReference? {
        guard AppDelegate.isFirebaseConfigured else {
            print("⚠️ [FIREBASE_CONN] Firebase not configured — skipping write session (\(reason))")
            return nil
        }
        queue.sync {
            idleCloseWork?.cancel()
            idleCloseWork = nil
            writeSessionCount += 1
            if writeSessionCount == 1 {
                Database.database().goOnline()
                print("🔌 [FIREBASE_CONN] goOnline (\(reason)) sessions=1")
            } else {
                print("🔌 [FIREBASE_CONN] reuse connection (\(reason)) sessions=\(writeSessionCount)")
            }
        }
        return Database.database().reference()
    }

    /// Closes the RTDB connection when the last in-flight write finishes (after a 1s idle).
    static func endWriteSession(reason: String) {
        guard AppDelegate.isFirebaseConfigured else { return }
        queue.async {
            writeSessionCount = max(0, writeSessionCount - 1)
            print("🔌 [FIREBASE_CONN] endWriteSession (\(reason)) sessions=\(writeSessionCount)")
            guard writeSessionCount == 0 else { return }
            idleCloseWork?.cancel()
            let work = DispatchWorkItem {
                if writeSessionCount == 0 {
                    Database.database().goOffline()
                    print("🔌 [FIREBASE_CONN] goOffline (idle after \(reason))")
                }
            }
            idleCloseWork = work
            queue.asyncAfter(deadline: .now() + idleCloseDelaySeconds, execute: work)
        }
    }

    /// After `FirebaseApp.configure()` the SDK is online. Disconnect until a write session starts.
    static func goIdleAfterConfigure() {
        guard AppDelegate.isFirebaseConfigured else { return }
        queue.async {
            if writeSessionCount == 0 {
                Database.database().goOffline()
                print("🔌 [FIREBASE_CONN] goOffline (post-configure idle)")
            }
        }
    }

    /// Spreads connection opens across launches using a stable per-device delay (0–20s).
    /// 200 devices launching together → ~10 starts/sec over 20s; short sessions aim for ~25 concurrent.
    static func jitterDelayMs(for userId: String, maxJitterMs: Int = launchJitterMaxMs) -> Int {
        guard !userId.isEmpty else { return 0 }
        let hash = abs(userId.hashValue)
        return hash % (maxJitterMs + 1)
    }

    /// Festival year for Firebase paths: always `Current::eventYear` from the pointer file,
    /// never UI browse year or calendar fallback. Returns 0 when unknown — callers must skip writes.
    static func firebaseStorageEventYear(maxWaitSeconds: TimeInterval = 0) -> Int {
        let deadline = Date().addingTimeInterval(maxWaitSeconds)
        repeat {
            if let year = pointerConfigCurrentEventYearInt(), year > 2000 {
                return year
            }

            var memoryYear: Int?
            storePointerLock.sync {
                if let value = cacheVariables.storePointerData["Current:eventYear"],
                   let year = Int(value.trimmingCharacters(in: .whitespacesAndNewlines)),
                   year > 2000 {
                    memoryYear = year
                }
            }
            if let year = memoryYear {
                return year
            }

            if maxWaitSeconds > 0, Date() < deadline {
                Thread.sleep(forTimeInterval: 0.5)
            } else {
                break
            }
        } while Date() < deadline

        print("❌ [FIREBASE_CONN] Unable to resolve pointer Current event year for Firebase storage")
        return 0
    }
}

/// Filter Xcode console with: `FIREBASE_SYNC_TRACE`
enum FirebaseSyncTrace {
    static let prefix = "[FIREBASE_SYNC_TRACE]"

    static func log(_ step: String, _ detail: String = "") {
        if detail.isEmpty {
            print("\(prefix) \(step)")
        } else {
            print("\(prefix) \(step) | \(detail)")
        }
    }

    static func snapshot(_ label: String) {
        let monitor = FirebaseWriteMonitor.shared
        let uid = UIDevice.current.identifierForVendor?.uuidString ?? "nil"
        let uidShort = uid.count > 8 ? String(uid.prefix(8)) + "…" : uid
        let storageYear = FirebaseConnectionHelper.firebaseStorageEventYear()
        let profile = SharedPreferencesManager.shared.getActivePreferenceSource()
        log(
            "SNAPSHOT \(label)",
            "uid=\(uidShort) profile=\(profile) bandDirty=\(monitor.hasPendingBandChanges()) showDirty=\(monitor.hasPendingShowChanges()) pendingFailures=\(monitor.hasPendingFailures()) shouldRunFullSync=\(monitor.shouldRunFullSync()) shouldRunBandSync=\(monitor.shouldRunBandSync()) firebaseConfigured=\(AppDelegate.isFirebaseConfigured) storageYear=\(storageYear) uiEventYear=\(eventYear) inTestEnvironment=\(inTestEnvironment)"
        )
    }
}
